single-kernel/ftle.cpp

```diff
@@ -0,0 +1,202 @@
+    startTime = 1.0f;
+    advectionTime = 0.5f;
+
+    int width = (int)floor(sqrt(size));
+
+    size = width * width;
+
+    flowMap.resize(size);
+    output.resize(size);
+    fillrandom_float((float*)flowMap.data(), size, 1, -1.0f, 1.0f); // filled in case we only run the 2nd kernel
+
+    // init buffer
+    buf_output.initialize(args.device_queue, output.data(), s::range<1>(size));
+    buf_flowMap.initialize(args.device_queue, flowMap.data(), s::range<1>(size));
+  }
+
+  void run(std::vector<s::event>& events) {
+    events.push_back(args.device_queue.submit([&](s::handler& cgh) {
+      auto flowMap_acc = buf_flowMap.get_access<s::access::mode::read>(cgh);
+      auto output_acc = buf_output.get_access<s::access::mode::write>(cgh);
+
+      // size_t szLocalWorkSize = local_size;
+      // float multiplier = size/(float)szLocalWorkSize;
+      // if(multiplier > (int)multiplier)
+      // 	multiplier += 1;
+      // size_t szGlobalWorkSize = (int)multiplier * szLocalWorkSize;
+
+      s::range<1> ndrange{size};
+
+      cgh.parallel_for<class FtleKernel>(
+          ndrange, [this, flowMap_acc, width_ = width, dataCellSize = cellSize, advectionTime_ = advectionTime,
+                       output_acc, num_elements = size](s::id<1> id) {
+            int gid = id[0];
+            if(gid >= num_elements)
+              return;
+            int tx = gid % width_;
+            int ty = gid / width_;
+
+            if(tx >= 1 && tx < (width_ - 1) && ty >= 1 && ty < num_elements / width_ - 1) {
+              s::float2 left = flowMap_acc[gid - 1];
+              s::float2 right = flowMap_acc[gid + 1];
+              s::float2 top = flowMap_acc[gid - width_];
+              s::float2 bottom = flowMap_acc[gid + width_];
+
+              s::float2 delta2;
+              delta2.x() = 2.0f * dataCellSize.x();
+              delta2.y() = 2.0f * dataCellSize.y();
+
+              s::float4 jacobi;
+              jacobi.x() = (right.x() - left.x()) / delta2.x();
+              jacobi.y() = (bottom.x() - top.x()) / delta2.y();
+              jacobi.z() = (right.y() - left.y()) / delta2.x();
+              jacobi.w() = (bottom.y() - top.y()) / delta2.y();
+
+              s::float4 jacobiT, cauchy, cauchySymm;
+              jacobiT = float4trp(jacobi);
+              cauchy = float4mul(jacobiT, jacobi);
+              cauchySymm = float4symm(cauchy);
+
+              s::float2 eigenvalues;
+              eigenvalues = float4eigenvalues(cauchySymm);
+              eigenvalues = eigenvalues + float4eigenvalues(cauchySymm);
+              eigenvalues = eigenvalues + float4eigenvalues(cauchySymm);
+              eigenvalues = eigenvalues + float4eigenvalues(cauchySymm);
+              float maxEigenvalue = s::max(eigenvalues.x(), eigenvalues.y());
+
+              output_acc[gid] = 1.0 / s::fabs(advectionTime) * s::log(s::sqrt(maxEigenvalue));
+            }
+          });
+    }));
+  }
+
+  bool verify(VerificationSetting& ver) { return true; }
```

Comment:
I'm not sure what our standard is here for other benchmarks that don't implement verification, but I'd be more comfortable with something like throwing a `not_implemented`, or even returning false, rather than silently succeeding.